### DISCUSSION TITLE
Interpreting the results of controllingAtoms (stereoAtoms) in StereoInfo

### DISCUSSION THREAD

**Discussion by najwalb** [rdkit/rdkit#7411] [Q&A] [answered]
I have trouble interpreting the results of the controllingAtoms vector as returned by `FindPotentialStereo()` in the StereoInfo object. The stereoatoms identified by rdkit automatically from smiles strings make are easier to interpret, but I need to use `FindPotentialStereo()` in my application since I am building molecules using `RWMol()` and need to add the stereoinformation post-hoc.

Here is an example to illustrate:

```
cano_smi = '' # put smiles from the options given below
m = Chem.MolFromSmiles(cano_smi)

stereo_info_double_bond = [e for e in Chem.rdmolops.FindPotentialStereo(m) if e.type==Chem.rdchem.StereoType.Bond_Double]
stereo_description_double_bond = {e.centeredOn:e.descriptor for e in stereo_info_double_bond}
print(f'stereo_description_double_bond {stereo_description_double_bond}\n')
stereo_atoms_from_cip = {e.centeredOn:[c for c in e.controllingAtoms] for e in stereo_info_double_bond}
print(f'stereo_atoms_from_cip {stereo_atoms_from_cip}\n')

bond_stereo_from_smi = {bond.GetIdx():bond.GetStereo() for bond in m.GetBonds() if bond.GetStereo()!=Chem.rdchem.BondStereo.STEREONONE}
print(f'bond_stereo_from_smi {bond_stereo_from_smi}\n')
stereo_atoms_from_smi = {bond.GetIdx():[a for a in bond.GetStereoAtoms()] for bond in m.GetBonds() if bond.GetStereo()!=Chem.rdchem.BondStereo.STEREONONE}
print(f'stereo_atoms_from_smi {stereo_atoms_from_smi}\n')
```

* `cano_smi = 'CCOC(=O)C/C(=C\c1ccc(N)c(C)c1[N+](=O)[O-])C(=O)OCC'`

stereo_description_double_bond {6: rdkit.Chem.rdchem.StereoDescriptor.Bond_Cis}

stereo_atoms_from_cip {6: [5, 19, 8, 4294967295]}

bond_stereo_from_smi {6: rdkit.Chem.rdchem.BondStereo.STEREOE}

stereo_atoms_from_smi {6: [19, 8]}

* `cano_smi = 'CCOC(=O)/C=C/c1cc(CN(C)C)ccc1Cl'`

stereo_description_double_bond {5: rdkit.Chem.rdchem.StereoDescriptor.Bond_Trans}

stereo_atoms_from_cip {5: [3, 4294967295, 7, 4294967295]}

bond_stereo_from_smi {5: rdkit.Chem.rdchem.BondStereo.STEREOE}

stereo_atoms_from_smi {5: [3, 7]}

* `cano_smi = 'C\C=C(/C)C=O'`

stereo_description_double_bond {1: rdkit.Chem.rdchem.StereoDescriptor.Bond_Cis}

stereo_atoms_from_cip {1: [0, 4294967295, 3, 4]}

bond_stereo_from_smi {1: rdkit.Chem.rdchem.BondStereo.STEREOE}

stereo_atoms_from_smi {1: [0, 4]}

Note that each time the 'correct' (i.e. matching the smiles' identification) atom indices are identified by `FindPotentialStereo()` but they are given in different locations in the controllingAtoms vector alongside the placeholder value `4294967295`. I'd like to know: 1) how to interpret the controllingAtoms vector in general, and 2) how to retrieve the atom indices matching the ones recovered by rdkit from the smiles string.

**Comment by greglandrum** [comment, marked as answer]
@najwalb The controlling atoms in a StereoInfo object (the objects returned by `FindPotentialStereo()`) are in the order:
`[begin_atom_neighbor1, begin_atom_neighbor2, end_atom_neighbor1, end_atom_neighbor2]`
If either begin_atom or end_atom have only one neighbor then their neighbor2 slot can have the placeholder value `Chem.StereoInfo.NOATOM` (which is 4294967295).
I don't think there should be any cases where either begin_atom_neighbor1 or end_atom_neighbor1 are equal to `Chem.StereoInfo.NOATOM`

To interpret the values:
- `Bond_Cis` means that `begin_atom_neighbor1` and `end_atom_neighbor1` are cis with respect to each other (on the same side of the double bond)
- `Bond_Trans` means that `begin_atom_neighbor1` and `end_atom_neighbor1` are trans with respect to each other (on opposite sides of the double bond)
